what4/test/Abduct.hs

```diff
@@ -0,0 +1,158 @@
+import What4.Solver
+import What4.Symbol (SolverSymbol(..))
+import What4.Protocol.SMTLib2 as SMT2
+         (assume, sessionWriter, runCheckSat, runGetAbducts, Writer)
+import What4.Protocol.SMTWriter
+         (mkSMTTerm)
+import What4.Protocol.Online
+
+cvc5executable :: FilePath
+cvc5executable = "cvc5"
+
+-- Call the online getAbduct tactic
+testGetAbductOnline ::
+  ExprBuilder t st fs ->
+  [BoolExpr t] ->
+  BoolExpr t ->
+  Int ->
+  IO [String]
+testGetAbductOnline sym hs g n = do
+  -- Print SMT file in /tmp/
+  withSystemTempFile "what4abdonline" $ \fname mirroredOutput -> do
+    proc <- startSolverProcess @(SMT2.Writer CVC5) cvc5Features (Just mirroredOutput) sym
+    let conn = solverConn proc
+    inNewFrame proc $ do
+      mapM_ (\x -> assume conn x) hs
+      getAbducts proc n (pack "abd") g
+    -- hClose mirroredOutput
+
+-- Call the offline getAbduct tactic
+testGetAbductOffline ::
+  ExprBuilder t st fs ->
+  BoolExpr t ->
+  Int ->
+  IO [String]
+testGetAbductOffline sym f n = do
+  -- Print SMT file in /tmp/
+  withSystemTempFile "what4abdoffline" $ \fname mirroredOutput -> do
+    let logData = LogData { logCallbackVerbose = \_ _ -> return ()
+                           , logVerbosity = 2
+                           , logReason = "defaultReason"
+                           , logHandle = Just mirroredOutput }
+    withCVC5 sym cvc5executable logData $ \session -> do
+      f_term <- mkSMTTerm (sessionWriter session) f
+      runGetAbducts session n (pack "abd") f_term
+    --hClose mirroredOutput
+
+-- Prove f using an SMT solver, by checking if ~f is unsatisfiable
+prove ::
+  ExprBuilder t st fs ->
+  BoolExpr t ->
+  [(String, IntegerExpr t)] ->
+  IO (SatResult () ())
+prove sym f es = do
+  -- Print SMT file in /tmp/
+  withSystemTempFile "what4prove" $ \fname mirroredOutput -> do
+    proc <- startSolverProcess @(SMT2.Writer CVC5) cvc5Features (Just mirroredOutput) sym
+    let logData = LogData { logCallbackVerbose = \_ _ -> return ()
+                           , logVerbosity = 2
+                           , logReason = "defaultReason"
+                           , logHandle = Just mirroredOutput }
+    
+    -- To prove f, we check whether not f is unsat
+    notf <- notPred sym f
+    withCVC5 sym cvc5executable logData $ \session -> do
+      checkSatisfiable proc "test" notf
+
+-- Tests
+
+testAbdOnline :: ExprBuilder t st fs -> 
+  [BoolExpr t] -> 
+  BoolExpr t -> 
+  TestTree
+testAbdOnline sym hs g = testCase "getting 3 abducts using cvc5 online" $ do
+  -- Ask for 3 abducts for f
+  res <- testGetAbductOnline sym hs g 3
+  (Data.List.length res == 3) @? "3 online abducts"
+
+testAbdOffline :: ExprBuilder t st fs -> 
+  BoolExpr t -> 
+  [(String, IntegerExpr t)] -> 
+  TestTree
+testAbdOffline sym f es = testCase "getting 3 abducts using cvc5 offline" $ do
+  -- Ask for 3 abducts for f
+  res <- testGetAbductOffline sym f 3
+  (Data.List.length res == 3) @? "3 offline abducts"
```

Comment:
Any reason not to just write `length` here instead of `Data.List.length`?

Reply:
I used it because of a name clash with `Data.Text.length`

Reply:
Ah, I overlooked that. Indeed, it's generally not advisable to import `Data.Text` unqualified, as its API overlaps with that of the `Prelude`. I'd recommend doing this instead:

```
import qualified Data.Text as Text
```

And then qualify all uses of `Text` functions (e.g., `Text.length`). That way, you shouldn't need to import `Data.List` at all.